solver/llbsolver/ops/diff.go

```diff
@@ -0,0 +1,105 @@
+package ops
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/moby/buildkit/worker"
+	"github.com/pkg/errors"
+
+	"github.com/moby/buildkit/cache"
+	"github.com/moby/buildkit/session"
+	"github.com/moby/buildkit/solver"
+	"github.com/moby/buildkit/solver/llbsolver"
+	"github.com/moby/buildkit/solver/pb"
+	digest "github.com/opencontainers/go-digest"
+)
+
+const diffCacheType = "buildkit.diff.v0"
+
+type diffOp struct {
+	op     *pb.DiffOp
+	worker worker.Worker
+}
+
+func NewDiffOp(v solver.Vertex, op *pb.Op_Diff, w worker.Worker) (solver.Op, error) {
+	if err := llbsolver.ValidateOp(&pb.Op{Op: op}); err != nil {
+		return nil, err
+	}
+	return &diffOp{
+		op:     op.Diff,
+		worker: w,
+	}, nil
+}
+
+func (d *diffOp) CacheMap(ctx context.Context, group session.Group, index int) (*solver.CacheMap, bool, error) {
+	dt, err := json.Marshal(struct {
+		Type string
+		Diff *pb.DiffOp
+	}{
+		Type: diffCacheType,
+		Diff: d.op,
+	})
+	if err != nil {
+		return nil, false, err
+	}
+
+	cm := &solver.CacheMap{
+		Digest: digest.Digest(dt),
+		Deps: make([]struct {
+			Selector          digest.Digest
+			ComputeDigestFunc solver.ResultBasedCacheFunc
+			PreprocessFunc    solver.PreprocessFunc
+		}, 2),
```

Comment:
2 -> should come from proto

Reply:
Updated to set this to the number of non-empty inputs (i.e. at most 2 if both lower and upper are non-empty). Let me know if that's what you were thinking.